osu.Desktop/GameStateBroadcaster.cs

```diff
@@ -0,0 +1,263 @@
+
+                    var client = new WebSocketClient(wsContext.WebSocket);
+                    client.OnStart += add;
+                    client.OnClose += remove;
+                    client.OnReady += () => client.Enqueue(getStateAsString());
+
+                    client.Start();
+                }
+                catch (HttpListenerException)
+                {
+                }
+            }
+        }
+
+        private void add(WebSocketClient client)
+        {
+            lock (clients)
+                clients.Add(client);
+        }
+
+        private void remove(WebSocketClient client)
+        {
+            lock (clients)
+                clients.Remove(client);
+        }
+
+        private void broadcast()
+        {
+            if (!enabled.Value)
+                return;
+
+            broadcastSchedule?.Cancel();
+            broadcastSchedule = Scheduler.AddDelayed(() =>
+            {
+                lock (clients)
+                {
+                    foreach (var client in clients)
+                        client.Enqueue(getStateAsString());
+                }
+            }, 200);
+        }
+
+        private void stop(bool closing = false)
+        {
+            cancellationToken?.Cancel();
+            broadcastSchedule?.Cancel();
+            Task.Run(async () =>
+            {
+                IEnumerable<Task> closeTasks;
+                lock (clients)
+                    closeTasks = clients.Select(c => c.Close());
+
+                await Task.WhenAll(closeTasks).ConfigureAwait(false);
+
+                if (listener.IsListening)
+                    listener.Stop();
+
+                if (closing)
+                    listener.Close();
+            });
+        }
+
+        private string getStateAsString()
+        {
+            return JsonConvert.SerializeObject(state, new JsonSerializerSettings
+            {
+                ReferenceLoopHandling = ReferenceLoopHandling.Ignore
+            });
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            base.Dispose(isDisposing);
+            stop(true);
+        }
+
+        private sealed class WebSocketClient
+        {
+            private readonly WebSocket socket;
+
+            private readonly CancellationTokenSource cancellationToken = new CancellationTokenSource();
+
+            private readonly BlockingCollection<string> queue = new BlockingCollection<string>();
+
+            public event Action<WebSocketClient> OnClose;
+
+            public event Action<WebSocketClient> OnStart;
+
+            public event Action OnReady;
+
+            private bool hasStarted;
+
+            public WebSocketClient(WebSocket socket)
+            {
+                this.socket = socket;
+            }
+
+            public void Start()
+            {
+                OnStart?.Invoke(this);
+                Task.WhenAll(receive(), send()).ConfigureAwait(false);
```

Comment:
Likewise, are you following some example for this implementation? If so, a reference would be appreciated (MSDN etc.)